File: c/src/mbug_2165.c

//==============================================================================
/* MBUG-2165 interface implementation

		Any changes in this file should also be applied in mbug_2151.c.
*/
//==============================================================================

#include <stdlib.h>
#include <string.h>
#include "mbug_2165.h"

//------------------------------------------------------------------------------
// Parameter numbers (keep private, may change in future revisions)

enum  mbug_2165_param {
	PARAM_START = 0x10,
	PARAM_RESET = 0xFF,
	PARAM_ITER  = 0x01,
	PARAM_CLOCK_DIV = 0x02,
	PARAM_MOD_CLOCK_DIV = 0x03,
	PARAM_BASE_CLOCK = 0x05,
	PARAM_MOD_CLOCK = 0x06,
	PARAM_BUSY = 0x70,
	PARAM_SEQ_LEN = 0x71
};

//------------------------------------------------------------------------------
// Device base and modulation clock.  These are assumed to be identical for
//   all mbug-2165 devices and only read the very first time a device is opened.
//   Currently, _base_clock = _mod_clock = 4 MHz.

static int base_clock = 0;
static int mod_clock = 0;

//------------------------------------------------------------------------------
// Get a list of all available mbug-2165 devices. 
const mbug_device_list  mbug_2165_list( void )
{
	return mbug_get_device_list(2165);
}

//------------------------------------------------------------------------------
// Open a device specified by it's serial number (as int,
// last digits of the serial number are matched only).
mbug_device mbug_2165_open( int serial_num )
{
	mbug_device dev = mbug_open_int(2165, serial_num);
	if ( (dev != NULL) && (base_clock == 0) )
	{
		base_clock = mbug_2165_get_base_clock( dev );
		mod_clock = mbug_2165_get_mod_clock( dev );
	}
	return dev;
}

//------------------------------------------------------------------------------
// Open a device specified by it's id string, as returned
// by mbug_2820_list()
mbug_device mbug_2165_open_str( const char *id )
{
	mbug_device dev;
	unsigned int type = mbug_type_from_id( id );
	if (type != 2165) return NULL;
	dev = mbug_open_str( id );
	if ( (dev != NULL) && (base_clock == 0) )
	{
		base_clock = mbug_2165_get_base_clock( dev );
		mod_clock = mbug_2165_get_mod_clock( dev );
	}
	return dev;
}

//------------------------------------------------------------------------------
// Close a previously opened device.
void mbug_2165_close( mbug_device dev )
{
	mbug_close(dev);
}

//------------------------------------------------------------------------------
// Private helper function:
//   Generic I/O command: Send cmd, receive integer parameter
//   Return values <0 indicate errors.
static int io( mbug_device dev,  unsigned char* out,  size_t size )
{
	unsigned char in[8] = {0};
	short status = 0;
	int value = 0;

	if (mbug_write( dev, out, size ) <0)
		return -1;
	if (mbug_read( dev, in, sizeof(in) ) <0)
		return -1;

	// First 2 bytes contain the return status (signed 16 bit)
	status = *(short*) in;
	// Next 4 bytes contain the returned output value
	value = *(int*) (in+2);
	if (status<0)  return status;
	else  return value;
}

//------------------------------------------------------------------------------
// Private helper function:
//   Set parameter value.
static int set_param( mbug_device dev, enum mbug_2165_param param, unsigned int value )
{
	unsigned char out[8] = {0xB0, param, 0, 0, 0, 0, 0, 0};
	memcpy( out+2, (void*)&value, 4 );
	return io( dev, out, 6);
}

//------------------------------------------------------------------------------
// Private helper function:
//   Get parameter value.
static int get_param( mbug_device dev, enum mbug_2165_param param )
{
	unsigned char out[2] = {0xC0, param};
	return io( dev, out, 2);
}

//------------------------------------------------------------------------------
// Set transmission sequence (max 255 bytes)
int mbug_2165_set_sequence( mbug_device dev, unsigned char* data, size_t nbytes, mbug_2165_tx_mode tx_mode )
{
	unsigned char out[64] = {0xB0, 0};
	unsigned char* pd = data;
	size_t  s, smax;

	if (nbytes>255)  nbytes=255;
	if (tx_mode > TX_MODE_TIMED_16)
		tx_mode = TX_MODE_BITSTREAM;

	smax = 64 - 4;  // Maximum data length per packet

	// First packet: Set sequence
	// Command format: {0xB0, 0, mode<<4, len, data[:] }
	s = nbytes>smax ? smax : nbytes;
	out[2] = tx_mode<<4;  out[3] = s;
	memcpy( out+4, pd, s );
	nbytes -= s;  pd += s;
	if (io( dev, out, s+4 ) <0)
		return -1;

	// Subsequent packeges: Append sequence
	// Command format: {0xB0, 0, 0x01, len, data[] }
	while (nbytes>0)
	{
		s = nbytes>smax ? smax : nbytes;
		out[2] = 0x01;  out[3] = s;
		memcpy( out+4, pd, s );
		nbytes -= s;  pd += s;
		if (io( dev, out, s+4 ) <0)
			return -1;
	}
	return 0;
}

//------------------------------------------------------------------------------
// Set transmission sequence with 16-bit values (max 127 items)
int mbug_2165_set_sequence_16bit( mbug_device dev, unsigned short* data, size_t nitems )
{
	return mbug_2165_set_sequence( dev, (unsigned char*)data, nitems*2, TX_MODE_TIMED_16 );
}

//------------------------------------------------------------------------------
// Get the size (in bytes) of the currently set sequence
int mbug_2165_get_seq_length( mbug_device dev )
{
	return get_param( dev, PARAM_SEQ_LEN );
}

//------------------------------------------------------------------------------
// Set number of transmission sequence iterations,
// 0: infinite, Maximum: 2**16-1
int mbug_2165_set_iterations( mbug_device dev, unsigned int niter )
{
	if (niter >= 1<<16 )  niter = (1<<16)-1;
	return set_param( dev, PARAM_ITER, niter );
}

//------------------------------------------------------------------------------
// Get the currently set number of transmission sequence iterations,
// 0: infinite
int mbug_2165_get_iterations( mbug_device dev )
{
	return get_param( dev, PARAM_ITER );
}

//------------------------------------------------------------------------------
// Get the devices base transmission clock frequency in Hz
int mbug_2165_get_base_clock( mbug_device dev )
{
	return get_param( dev, PARAM_BASE_CLOCK );
}

//------------------------------------------------------------------------------
// Set clock divider. => Bitrate = base_clock/div
// Minimum div is 10 for bitstream mode, 1 for timed modes,
// Maximum div is 2**16-1 (65535).
int mbug_2165_set_clock_div( mbug_device dev, unsigned int div )
{
	if (div == 0)  div = 1;
	if (div >= 1<<16 )  div = (1<<16)-1;
	return set_param( dev, PARAM_CLOCK_DIV, div );
}

//------------------------------------------------------------------------------
// Get the currently set transmission clock divider.
int mbug_2165_get_clock_div( mbug_device dev )
{
	return get_param( dev, PARAM_CLOCK_DIV );
}

//------------------------------------------------------------------------------
// Set the transmission rate (via clock div).
// Minimum rate is 61Hz.
// Maximum rate is 400kHz for bitstream mode, 4Mhz for timed modes.
// Returns the actually set bitrate or <0 in case of an error.
double mbug_2165_set_bitrate( mbug_device dev, double freq )
{
	int div;
	div = (int)(0.5 + 1.0*base_clock/freq);
	if (div >= 1<<16)  div = (1<<16)-1;
	if( mbug_2165_set_clock_div( dev, div ) < 0)
		return -1;
	return (int)(0.5 + 1.0*base_clock/div);
}

//------------------------------------------------------------------------------
// Set the transmission timebase (in seconds, via clock div).
// Minimum time is 2.5us for bitstream mode, 0.25us for timed modes.
// Maximum time is 16.38375ms.
// Resolution is 0.25us.
// Returns the actually time or <0 in case of an error.
double mbug_2165_set_timebase( mbug_device dev, double timebase )
{
	int div;
	div = (int)(0.5 + 1.0*timebase*base_clock);
	if (div >= 1<<16)  div = (1<<16)-1;
	if( mbug_2165_set_clock_div( dev, div ) < 0)
		return -1;
	return (int)(0.5 + 1.0*div/base_clock);
}

//------------------------------------------------------------------------------
// Get the devices modulation base clock frequency in Hz
int mbug_2165_get_mod_clock( mbug_device dev )
{
	return get_param( dev, PARAM_MOD_CLOCK );
}

//------------------------------------------------------------------------------
// Set modulation clock divider. => mod_freq = base_clock/div
// Minimum modulation div is 16,
// Maximum modulation div is 2**12-1 (4095).
// div = 0: Turned OFF modulation.
int mbug_2165_set_mod_clock_div( mbug_device dev, unsigned int div )
{
	if (div >= 1<<16 )  div = (1<<16)-1;
	return set_param( dev, PARAM_MOD_CLOCK_DIV, div );
}

//------------------------------------------------------------------------------
// Get the currently set modulation clock divider.
int mbug_2165_get_mod_clock_div( mbug_device dev )
{
	return get_param( dev, PARAM_MOD_CLOCK_DIV );
}

//------------------------------------------------------------------------------
// Set the modulation frequency (via clock div).
// Minimum modulation frequency is 976.8 Hz, maximum is 250 kHz.
// Frequency == 0: Turn modulation off
// Returns the actually set frequency or <0 in case of an error.
double mbug_2165_set_mod_freq( mbug_device dev, double freq )
{
	int div;
	if (freq<0)
		freq=-freq;
	if (freq>0) {
		div = (int)(0.5 + 1.0*mod_clock/freq);
		if (div >= 1<<12)  div = (1<<12)-1;
		if (div < 16 )  div = 16;
	} else {
		div = 0;
	}
	if( mbug_2165_set_mod_clock_div( dev, div ) < 0)
		return -1;
	return div==0 ? 0 : (int)(0.5 + 1.0*mod_clock/div);
}

//------------------------------------------------------------------------------
// Start transmission.
int mbug_2165_start( mbug_device dev )
{
	return set_param( dev, PARAM_START, 0x0001 );
}


//------------------------------------------------------------------------------
// Stop transmission
//	 force = 0: Stop gracefully after ongoing iteration
//	 force = 1: Stop instantly
int mbug_2165_stop( mbug_device dev, int force )
{
	return set_param( dev, PARAM_START, force ? 0x0100 : 0x0000  );
}

//------------------------------------------------------------------------------
// Stop transmission gracefully, i.e. after the current iteration.
int mbug_2165_stop_gracefully( mbug_device dev )
{
	return set_param( dev, PARAM_START, 0x0000 );
}

//------------------------------------------------------------------------------
// Stop transmission instantly.
int mbug_2165_stop_immediately( mbug_device dev )
{
	return set_param( dev, PARAM_START, 0x0100 );
}

//------------------------------------------------------------------------------
// Reset transmission and all parameters
int mbug_2165_reset( mbug_device dev )
{
	return set_param( dev, PARAM_RESET, 0 );
}

//------------------------------------------------------------------------------
// Get busy state. >0 if transmission is in progress
int mbug_2165_get_busy( mbug_device dev )
{
	return get_param( dev, PARAM_BUSY );
}

//------------------------------------------------------------------------------

File: py/mbug_dev/mbug_2165.py

from mbug_base import *

#==========================================================================
class mbug_2165(mbug):
    """ Represents an MBUG-2165 IR transmitter"""
    
    _type = 2165

    def __init__(self, serial=None):
        mbug.__init__(self, self._type, serial)
        self._base_clock = self.get_base_clock()
        self._mod_base_clock = self.get_mod_base_clock()
        self.target._dev = self

    def _iocmd(self, cmd):
        self._write( cmd )
        ret = self._read()
        status = (ret[0] + ret[1]*256) - 2**16*(ret[1]&0x80>0)   # signed 16bit
        if status!=0:
            errmsg = {-1: 'Unknown command',
                }.get( status, 'Undefined error' )
            raise RuntimeError('mbug_2165: Communication Error: %s (%d)' % (errmsg,status) )
        return ret[2:]

    def set_sequence_bitstream(self, data):
        """ Set a transmission byte sequence in bitstream mode. The data will be 
        transmitted bit per bit, starting with the least significant bit of the 
        first byte. The maximum length is 255 bytes at the moment. """
        nbytes = len(data)
        if nbytes>255: raise ValueError('Maximum sequence length is 255.')
        s = self._size_out-4 # max data size per packet
        # First packet: set_sequence
        d, data = data[:s], data[s:]
        self._iocmd( bytearray(( 0xB0, 0x00, 0x00, len(d) )) + bytearray(d) )
        #Subsequent packets: append_sequence
        while (len(data)>0):
            d, data = data[:s], data[s:]
            self._iocmd( bytearray(( 0xB0, 0x00, 0x01, len(d) )) + bytearray(d) )
        return
        
    def set_sequence_times(self, data, unit='ms'):
        """Set a transmission sequence in timed mode. The sequence is interpreted
        as consecutive on/off intervalls (starting with on). The unit is determined
        by the unit parameter, either numerically (in seconds) or as 's' (seconds),
        'ms' (milliseconds), us (microseconds). Default unit is 'ms'.
        The resolution of the sequence times (determined by the time base tau and the
        word length (8 or 16 bits)) is chosen automatically.
        If you want more control over the resolution, use the more specific transmis-
        sion functions set_sequence_8bit() or set_sequence_16bit() together with
        set_tau() or set_clock_div()."""
        if isinstance(unit, int) or isinstance(unit,float): pass
        else:
            try: unit = {'us':1e-6, 'ms':1e-3, 's':1 }[unit]
            except KeyError: KeyError('Invalid unit: '+str(unit))
        for i,d in enumerate(data):
            data[i] *= unit
            if data[i]<40e-6: data[i]=40e-6
        nitems = len(data)
        if nitems>255: raise ValueError('Maximum sequence length is 255.')
        bits = 8 if nitems>127 else 16
        tau = max(data)/2**bits
        div = int(tau*self._base_clock + 1) # NOTE: Take ceiling! (Else data items may overflow.)
        if div >= 2**16:
            div = 2**16-1
            "### Warning: Some sequence items are too long and will be clipped. ###"
        tau = 1.*div/self._base_clock
        if min(data)/tau<10:
            print "### Warning: Some sequence items are too short and will have a low resolution. ###"
        for i,d in enumerate(data):
            data[i] = int(data[i]/tau + 0.5)
        self.set_tau(tau)
        if bits==8: self.set_sequence_times_8bit(data)
        else: self.set_sequence_times_16bit(data)
        return

    def set_sequence_times_8bit(self, data):
        """Set a transmission sequence in timed 8 bit mode. The sequence is
        interpreted as consecutive on/off intervalls (starting with on) in units
        of the timebase tau (set via set_tau() or set_clock_div()).
        The size of a sequence element is limited to 8 bit (<=255). The sequence
        length is limited to a mximum of 255 elements."""
        nbytes = len(data)
        if nbytes>255: raise ValueError('Maximum sequence length is 255.')
        s = self._size_out-4 # max data size per packet
        # First packet: set_sequence
        d, data = data[:s], data[s:]
        self._iocmd( bytearray(( 0xB0, 0x00, 0x10, len(d) )) + bytearray(d) )
        #Subsequent packets: append_sequence
        while (len(data)>0):
            d, data = data[:s], data[s:]
            self._iocmd( bytearray(( 0xB0, 0x00, 0x01, len(d) )) + bytearray(d) )
        return

    def set_sequence_times_16bit(self, data):
        """Set a transmission sequence in timed 16 bit mode. The sequence is
        interpreted as consecutive on/off intervalls (starting with on) in units
        of the timebase tau (set via set_tau() or set_clock_div()).
        The size of a sequence element is limited to 16 bit (<=65535). The sequence
        length is limited to a mximum of 255 elements."""
        nitems = len(data)
        if nitems>127: raise ValueError('Maximum sequence length is 127.')
        bytes = []
        for item in data: bytes += [ item&0xFF, (item>>8)&0xFF]
        s = self._size_out-4 # max data size per packet
        # First packet: set_sequence
        d, bytes = bytes[:s], bytes[s:]
        self._iocmd( bytearray(( 0xB0, 0x00, 0x20, len(d) )) + bytearray(d) )
        #Subsequent packets: append_sequence
        while (len(bytes)>0):
            d, bytes = bytes[:s], bytes[s:]
            self._iocmd( bytearray(( 0xB0, 0x00, 0x01, len(d) )) + bytearray(d) )
        return
    
    def set_iterations(self, niter):
        """ Set the number of transmission iterations.
        0: infinite, Maximum: 2**16-1 at the moment. """
        if niter=='inf': niter=0
        if niter<0 or niter>=1<<16:
            raise ValueError
        self._iocmd( (0xB0, 0x01, niter&0xFF, (niter>>8)&0xFF, 0x00, 0x00 ) )
        return

    def set_clock_div(self, div):
        """ Set the clock divider. Bitrate = base_clock/div, tau = base_time*div
        Minimum div in bitstream mode is 10, maximum div is 2**16-1. """
        div = int(div)
        if div<=0: div=1
        if div>=1<<16: div=(1<<16)-1
        self._iocmd( (0xB0, 0x02, div&0xFF, (div>>8)&0xFF, 0x00, 0x00 ) )
        return

    def set_bitrate(self, freq):
        """ Set the bit rate (via clock div)."""
        if freq<=0: raise ValueError('Invalid bitrate.')
        div = round(self._base_clock/freq)
        self.set_clock_div(div)

    def set_timebase(self, tau):
        """ Set the timebase (via clock div)."""
        if tau<=0: raise ValueError('Invalid timebase')
        div = round(self._base_clock*tau)
        self.set_clock_div(div)
        
    def set_mod_clock_div(self, div):
        """ Set the modulation clock divider. freq = mod_base_clock/div.
        Minimum div is 16, maximum div is 4096 at the moment.
        Resolution decreases for div>256.
        Set div=0 to turn off modulation."""
        div = int(div)
        if div!=0 and (div<=0 or div>4096):
            raise ValueError('Invalid clock divider.')
        self._iocmd( (0xB0, 0x03, div&0xFF, (div>>8)&0xFF, 0x00, 0x00 ) )
        return

    def set_mod_freq(self, freq):
        """ Set the modulation frequency (via modulation clock div).
        Set freq=0 to turn off modulation."""
        if freq==0:
            div = 0
        else:
            div = round(self._mod_base_clock/freq)
            if div<16: div=16
            if div>4096: div=4096
            rfreq = self._mod_base_clock/div
            rd = 1.*rfreq/freq-1.
            if abs(rd)>0.01:
                print "Warning. Set modulation frequency to %.1e."%(rfreq)
        self.set_mod_clock_div(div)

    def start(self):
        """ Start transmission. """
        self._iocmd( (0xB0, 0x10, 0x01) )
        return

    def stop(self, instantly=0):
        """ Stop transmission.
        instantly==0: Stop gracefully after the current iteration.
        instantly==1: Stop instantly. (Discards remaining sequence bytes.)"""
        if instantly: self._iocmd( (0xB0, 0x10, 0x00, 0x01) )
        else:         self._iocmd( (0xB0, 0x10, 0x00, 0x00) )
        return

    def reset(self):
        """ Reset transmission and all parameters. """
        self._iocmd( (0xB0, 0xFF) )
        return

    def get_iterations(self):
        """ Query the number of transmission iterations (0:infinite)."""
        r = self._iocmd( (0xC0, 0x01) )   # Query base clock
        return sum( r[i]<<(i*8) for i in range(4) )

    def get_clock_div(self):
        """ Get the bit clock divider. Bitrate = base_clock/div."""
        r = self._iocmd( (0xC0, 0x02) )   # Query base clock
        return sum( r[i]<<(i*8) for i in range(4) )

    def get_mod_clock_div(self):
        """ Get the modulation clock divider. mod_freq = mod_base_clock/mod_div."""
        r = self._iocmd( (0xC0, 0x03) )   # Query base clock
        return sum( r[i]<<(i*8) for i in range(4) )
    
    def get_base_clock(self):
        """ Query the devices base bit clock frequency in Hz. """
        r = self._iocmd( (0xC0, 0x05) )   # Query base clock
        return sum( r[i]<<(i*8) for i in range(4) )

    def get_mod_base_clock(self):
        """ Query the devices modulation base clock frequency in Hz. """
        r = self._iocmd( (0xC0, 0x06) )   # Query base clock
        return sum( r[i]<<(i*8) for i in range(4) )
        
    def get_busy(self):
        """ Get the busy state. True if transmission is in progress. """      
        r = self._iocmd( (0xC0, 0x70) )   # Query tx_busy
        return bool( r[0] )

    def get_sequence_length(self):
        """ Get the size of the stored sequence in bytes """
        r = self._iocmd( (0xC0, 0x71) )  # Query seq_length
        return int( r[0]+r[1]*256 )
    
#----------------------------------------------------------------------------

    class target:
        """ A collection of target devices addressable by the mbug_2151.
        The target device must be created from an mbug_2151 instance, which
        will be used as controller."""
        
        _dev = None     # The controller device

        # Target device factories
        def Nikon_L3(self):
            return Nikon_L3(dev=self._dev)
        def Pentax(self):
            return Pentax(dev=self._dev)
        def Canon_RC1(self):
            return Canon_RC1(dev=self._dev)
        def Canon_WLD5000(self):
            return Canon_WLD5000(dev=self._dev)
        
    target = target()   # Make this an instance 


#==========================================================================

import time as _time

#==========================================================================

class mbug_2165_target(object):
    _dev = None          # Controller device
    _bitrate = None       # Default parameters
    _mod_freq = None
    _iterations = None
    _timeout = None

    def __init__(self, dev):
        if hasattr(dev,'_type') and dev._type==2165:
            self._dev = dev
        else:
            raise Exception('mbug_2165_target.init(): Need an mbug_2165 instance as control device.')

    def _wait_busy(self, force=0):
        # Wait for device to become ready
        if self._dev.get_busy():
            if not force:
                tout = _time.time() + self._timeout
                while (self._dev.get_busy()):
                    _time.sleep(0.1)
                    if _time.time()>tout:
                        break
            self._dev.disable()

    def _send(self, sequence, force=0):
        # Wait for device to become ready
        self._wait_busy(force)
        # Resend parameters since the device may have been used for another target in the meantime
        self._dev.set_bitrate(self._bitrate)
        self._dev.set_iterations(self._iterations)
        self._dev.set_mod_freq(self._mod_freq)
        self._dev.set_sequence_bitstream(sequence)
        # Engage!
        self._dev.start()
        
#---------------------------------------------------------------------------

class Nikon_L3(mbug_2165_target):
    """Represents a Nikon ML-L3 compatible camera."""

    # Transmission parameters
    _bitrate = 2507.837  # -> tau=1/(4e6/1595)=0.39875 ms
    _iterations = 2
    _mod_freq = 38400
    _timeout = 1
    _seq_trigger = [0b00011111, 0, 0, 0, 0, 0, 0, 0, 0, 0b00001000, 0b00000001, 0b00000100, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

    def __init__(self, dev):
        mbug_2165_target.__init__(self, dev)

    def trigger(self, force=0):
        self._send( self._seq_trigger, force )

#---------------------------------------------------------------------------

class Pentax(mbug_2165_target):
    """Represents a Pentax compatible camera."""

    # Transmission parameters
    _bitrate = 1e3  # -> tau=1 ms
    _iterations = 1
    _mod_freq = 38e3
    _timeout = 1
    _seq_trigger = [0b11111111, 0b00011111, 0b01010101, 0b00010101]

    def __init__(self, dev):
        mbug_2165_target.__init__(self, dev)

    def trigger(self, force=0):
        self._send( self._seq_trigger, force )

#---------------------------------------------------------------------------

class Canon_RC1(mbug_2165_target):
    """Represents a Canon RC-1 compatible camera."""

    # Transmission parameters
    _bitrate = 8333  # -> tau=0.12 ms
    _iterations = 1
    _mod_freq = 32.76e3
    _timeout = 1
    _seq_trigger = [0b00001111, 0, 0, 0, 0, 0, 0, 0 ]*2
    _seq_trigger_3s = [0b00001111, 0, 0, 0, 0, 0 ]*2
    
    def __init__(self, dev):
        mbug_2165_target.__init__(self, dev)

    def trigger(self, force=0):
        self._send( self._seq_trigger, force )

    def trigger_3s(self, force=0):
        self._send( self._seq_trigger_3s, force )

#---------------------------------------------------------------------------

class Canon_WLD5000(mbug_2165_target):
    """Canon Video camera remote control WL-D5000"""
    
    # similar to NEC protocol,
    #  - no repeat shortcut (message is repeated in full length)
    #  - 2nd addr field inversion is not obeyed
    
    _bitrate = 7142  # -> tau=0.14 ms, 1.12 ms per byte
    _iterations = 3
    _mod_freq = 38e3
    _timeout = 1
    _1_, _0_ = [0x0f,0], [0x0f]
    _seq_trigger = [0xff]*8 + [0]*4 \
                   + _1_+_0_+_1_+_0_+_0_+_0_+_0_+_1_ + _0_+_1_+_1_+_0_+_1_+_1_+_1_+_0_ \
                   + _0_+_0_+_0_+_0_+_1_+_1_+_0_+_0_ + _1_+_1_+_1_+_1_+_0_+_0_+_1_+_1_ \
                   + _0_ + [0]*36

## Timed mode:
##    _seq_trigger = [9.00, 4.45] + [ (0.59,1.65)[b] for b in \
##        (1,0,1,0,0,0,0,1,  0,1,1,0,1,1,1,0,  0,0,0,0,1,1,0,0,  1,1,1,1,0,0,1,1) ] 
    
    def __init__(self, dev):
        mbug_2165_target.__init__(self, dev)

    def trigger(self, force=0):
        self._send( self._seq_trigger, force )

#==========================================================================

def mbug_2165_test():
    # Print device list
    devs = list_devices(2165)
    print "Attached MBUG-2165 devices:"
    print devs if devs!=[] else 'None'

    # Open devices
    for ser in devs:
        try:
            print "Device", ser
            dev = mbug_2165(ser)
            ##tar = mbug_2165.target.Canon_WLD5000()
            ##tar.trigger()
            dev.set_sequence_times([0.1,0.1,0.2,0.2,0.5,0.5,1,1,2,2,5,5,10,10,20,20,50,50,100,100])
            ##dev.set_iterations(3)
            ##dev.set_clock_div(4)
            ##dev.set_sequence_times_16bit([100,100,200,200,500,500,1000,1000,10000,10000])
            ##dev.set_sequence_times([100,100,200,200,500,500,1000,1000,10000,10000],'us')
            ##dev.set_sequence_times([0.1,0.1,0.2,0.2,0.5,0.5,2,2,5,5,10,10,20,20,50,50,100,100],'ms')
            ##dev.set_sequence_times([0.1,0.1,0.2,0.2,0.5,0.5,2,2,5,5,10,10,20,20],'ms')
            dev.start()
        finally:
            try: dev.close()
            except: pass

#---------------------------------------------------------------------------
if __name__ == "__main__":
    pass
    mbug_2165_test()
        
#---------------------------------------------------------------------------
